Frameworks/AddressBook/ABAddressBookManagerInternal.mm

```diff
@@ -82,4 +192,183 @@ - (NSArray*)getListOfContacts {
+}
+
+- (bool)removeContact:(ABRecordRef)record error:(CFErrorRef*)error {
+    _ABContact* contact = (__bridge _ABContact*)record;
+
+    switch (contact.type) {
+        case kAddressBookNewContact: {
+            // New contacts can only be removed if they were
+            // previously marked to be added.
+            if ([self->_toAdd containsObject:contact.contact]) {
+                [self->_toAdd removeObject:contact.contact];
+                return true;
+            }
+
+            [self _setError:error message:@"Only existing contacts can be removed from the AddressBook.\n"];
+            return false;
+        }
+        case kAddressBookReadWriteContact: {
+            __ABContactOperation* operation = self->_operations[contact.contact.id];
+
+            // A read/write contact can be deleted in 2 cases: if no operation was associated,
+            // or if the associated operation is that it has been modified. In the case that
+            // it is already marked to delete, deleting twice doesn't make sense.
+            if (operation == nil) {
+                operation = [[__ABContactOperation alloc] initWithContact:contact.contact shouldDelete:YES];
+                self->_operations[contact.contact.id] = operation;
+                return true;
+            } else if (operation.shouldDelete) {
+                [self _setError:error message:@"This contact has already been deleted.\n"];
+                return false;
+            } else {
+                operation.shouldDelete = YES;
+                return true;
+            }
+        }
+        case kAddressBookReadOnlyContact:
+        default:
+            [self _setError:error
+                    message:@"Read-only contacts cannot be deleted. Use ABAddressBookCopyArrayOfAllUserAppPeople to get contacts that can "
+                            @"be deleted.\n"];
+            return false;
+    }
+}
+
+- (void)modifyContact:(ABRecordRef)record {
+    _ABContact* contact = (__bridge _ABContact*)record;
+    __ABContactOperation* operation = self->_operations[contact.contact.id];
+
+    // If there was no operation previously associated with this record, then
+    // set it to be modified. The other two cases are that the contact was
+    // already to be modified (in which case, we don't need to do anything),
+    // or that the contact is marked to be deleted -- in this case, since
+    // it is to be deleted, there is no need to worry about any modifications.
+    if (operation == nil) {
+        operation = [[__ABContactOperation alloc] initWithContact:contact.contact shouldDelete:NO];
+        self->_operations[contact.contact.id] = operation;
+    }
+}
+
+- (bool)hasUnsavedChanges {
+    return (self->_toAdd).count + (self->_operations).count > 0;
+}
+
+- (void)revert {
+    [self->_toAdd removeAllObjects];
+    [self->_operations removeAllObjects];
+}
+
+- (bool)save {
+    // TODO: When the projections for WACContactList's saveContactAsync and deleteContactAsync
```

Comment:
>TODO [](start = 7, length = 4)

Is this TODO associated with another task? If so, place the task's issue ID here for proper tracking. If not, please create one. :)